cs/codec-template.cs.j2

```diff
@@ -0,0 +1,269 @@
+
+        public static ClientMessage Encode{{ event.name|capital }}Event({% for param in event.params %}{{ lang_types_encode(param.type) }} {{param.name}}{% if not loop.last %}, {% endif %}{% endfor %})
+        {
+            var clientMessage = CreateForEncode();
+            var initialFrame = new Frame(new byte[Event{{ event.name|capital}}InitialFrameSize], UnfragmentedMessage);
+            initialFrame.Flags |= IsEventFlag;
+            EncodeInt(initialFrame.Content, TypeFieldOffset, Event{{ (event.name|capital)}}MessageType);
+        {% for param in fixed_params(event.params) %}
+            Encode{{ lang_types_decode(param.type)|capital }}(initialFrame.Content, Event{{ (event.name|capital)}}{{(param.name|capital)}}FieldOffset, {{ param.name }});
+        {% endfor %}
+            clientMessage.Add(initialFrame);
+        {% for param in var_size_params(event.params) %}
+            {{ encode_var_sized(param) }};
+        {% endfor %}
+            return clientMessage;
+        }
+        {% endfor %}
+
+        public static class EventHandler
+        {
+            public static void HandleEvent(ClientMessage clientMessage{% for event in method.events%}, Handle{{ event.name|capital }}Event handle{{ event.name|capital }}Event{% endfor %})
+            {
+                var messageType = clientMessage.MessageType;
+                var iterator = clientMessage.GetIterator();
+            {% for event in method.events%}
+                if (messageType == Event{{(event.name|capital)}}MessageType) {
+                {% if  fixed_params(event.params)|length != 0 %}
+                    var initialFrame = iterator.Next();
+                {% else %}
+                    //empty initial frame
+                    iterator.Next();
+                {% endif %}
+                {% for param in fixed_params(event.params) %}
+                    {{ lang_types_decode(param.type) }} {{param.name}} =  Decode{{ lang_types_decode(param.type)|capital }}(initialFrame.Content, Event{{(event.name|capital)}}{{param.name|capital}}FieldOffset);
+                {% endfor %}
+                {% for param in var_size_params(event.params) %}
+                    {{ lang_types_decode(param.type) }} {{param.name}} = {{ decode_var_sized(param) }};
+                {% endfor %}
+                    handle{{ event.name|capital }}Event({% for param in event.params %}{{param.name}}{% if not loop.last %}, {% endif %}{% endfor %});
+                    return;
+                }
+            {% endfor %}
+                Logger.GetLogger(typeof(EventHandler)).Finest("Unknown message type received on event handler :" + messageType);
+            }
+        {% for event in method.events%}
+            public delegate void Handle{{ event.name|capital }}Event({% for param in event.params %}{{ lang_types_encode(param.type) }} {{param.name}}{% if not loop.last %}, {% endif %}{% endfor %});
+        {% endfor %}
+        }
+    {% endif %}
+    }
+}
```

Comment:
Will there be a codec hash test for c# client ?

Reply:
I think we don't need.